deduction_rules.py

```diff
@@ -167,4 +167,162 @@ def line_square(sudoku):
+                    l.append(tuple(sorted((comb[0],comb[1],comb[2]))))
+                    break
+                elif len(cell0) == 3 and cell0 == cell1 and len(cell2) == 2 and cell2.issubset(cell1):
+                    l.append(tuple(sorted((comb[0],comb[1],comb[2]))))
+                    break
+                elif len(cell0) == 3 and len(cell1) == 2 and len(cell2) == 2 and \
+                    cell1.issubset(cell0) and cell2.issubset(cell0) and len(cell1.intersection(cell2)) == 1:
+                    l.append(tuple(sorted((comb[0],comb[1],comb[2]))))
+                    break
+                elif len(cell0) == 2 and len(cell1) == 2 and len(cell1) == 2 and \
+                    len(set.intersection(cell0,cell1)) == 2 and len(set.intersection(cell1,cell2)) == 2 and \
+                    len(set.intersection(cell0,cell2)) == 2 and len(set.intersection(cell0,cell1,cell2)) == 1:
+                    l.append(tuple(sorted((comb[0],comb[1],comb[2]))))
+                    break
+        return l
+    
+    def search_and_ban_in_subset(cells_to_check):
+        """Searches all naked-trios in a subset of cells and bans these numbers from the other elements of subset."""
+        made_deduction = False
+        allowed_numbers = _allowed_numbers(sudoku,cells_to_check)
+        naked_trios = search_for_naked_trios(allowed_numbers)
+        for trio in naked_trios:
+            print("naked_trios :)")
+            cell0 = cells_to_check[trio[0]]
+            cell1 = cells_to_check[trio[1]]
+            cell2 = cells_to_check[trio[2]]
+            deleted_numbers = set(sudoku.allowed[cell0[0]][cell0[1]].allowed() + sudoku.allowed[cell1[0]][cell1[1]].allowed() + sudoku.allowed[cell2[0]][cell2[1]].allowed())
+            cells_used = sudoku.allowed[cell0[0]][cell0[1]].notNones()+sudoku.allowed[cell1[0]][cell1[1]].notNones()+sudoku.allowed[cell2[0]][cell2[1]].notNones()
+            for cell in cells_to_check:
+                if cell not in (cell0, cell1, cell2):
+                    made_deduction |= _ban_numbers(sudoku,cell,deleted_numbers,"naked-trios",cells_used)
+        return made_deduction
+
+    return _apply_for_nines(search_and_ban_in_subset)
+
+def hidden_trios(sudoku):
+    """RULE: if only A,B,C cells (in one row/col/sec) has a number x,y,z then delete every number from A,B,C except x,y,z."""
+    def search_hidden_trios(elems):
+        res = dict()
+        where_can_go = {i:[] for i in range(1,10)} # i-th number in which indices of elems can go
+        for idx,elem in enumerate(elems):
+            for num in elem:
+                where_can_go[num].append(idx)
+        
+        for nums in combinations(range(1,10),3):
+            nums_can_go = tuple(set.union(set(where_can_go[nums[0]]), set(where_can_go[nums[1]]), set(where_can_go[nums[2]])))
+            if len(set(where_can_go[nums[0]])) > 1 and len(set(where_can_go[nums[1]])) > 1 and len(set(where_can_go[nums[2]])) > 1 and\
```

Comment:
All these `set` conversions make me wonder, why don't we define `where_can_go` as a `dict` of `set`s?

Reply:
fixed